### PR TITLE
Сделал лабораторную работу

### PR DESCRIPTION
**Pull request by nekitlnp**

### CODE REVIEW

**Comment by msmirnov-c** [comment]
1:4   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   2:55  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   3:81  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   4:3   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   5:10  error  'addNumber' is defined but never used                                                                                                                                  no-unused-vars
   5:36  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   6:8   error  There should be no spaces inside this paren                                                                                                                            space-in-parens
   6:21  error  'isSomeArrayValueNaN' was used before it was defined                                                                                                                   no-use-before-define
   6:42  error  There should be no spaces inside this paren                                                                                                                            space-in-parens
   6:45  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   7:1   error  Expected indentation of 8 spaces but found 4                                                                                                                           indent
   7:17  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   8:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
   9:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
   9:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  10:8   error  There should be no spaces inside this paren                                                                                                                            space-in-parens
  10:11  error  'isNumberPositiveAndInteger' was used before it was defined                                                                                                            no-use-before-define
  10:46  error  There should be no spaces inside this paren                                                                                                                            space-in-parens
  10:49  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  11:1   error  Expected indentation of 8 spaces but found 4                                                                                                                           indent
  11:5   error  Function 'addNumber' expected a return value                                                                                                                           consistent-return
  11:12  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  12:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  13:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  13:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  14:24  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  15:18  error  There should be no spaces inside this paren                                                                                                                            space-in-parens
  15:20  error  Expected parentheses around arrow function argument                                                                                                                    arrow-parens
  15:29  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  16:1   error  Expected indentation of 8 spaces but found 4                                                                                                                           indent
  16:34  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  17:8   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  18:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  18:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  19:50  error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  19:51  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  20:35  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  21:9   error  'outputArray' is never reassigned. Use 'const' instead                                                                                                                 prefer-const
  21:26  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  22:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  22:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  23:5   error  for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array  no-restricted-syntax
  23:5   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype                                                             guard-for-in
  23:14  error  'i' is never reassigned. Use 'const' instead                                                                                                                           prefer-const
  23:28  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  24:1   error  Expected indentation of 8 spaces but found 4                                                                                                                           indent
  24:47  error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  24:48  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  25:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  26:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  26:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  27:24  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  28:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  28:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  29:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  29:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  30:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  30:41  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  31:12  error  'isNotANumber' was used before it was defined                                                                                                                          no-use-before-define
  31:31  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  32:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  32:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  33:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  33:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  34:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  34:47  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  35:8   error  There should be no spaces inside this paren                                                                                                                            space-in-parens
  35:11  error  'isNotANumber' was used before it was defined                                                                                                                          no-use-before-define
  35:32  error  'isInteger' was used before it was defined                                                                                                                             no-use-before-define
  35:61  error  There should be no spaces inside this paren                                                                                                                            space-in-parens
  35:64  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  36:1   error  Expected indentation of 8 spaces but found 4                                                                                                                           indent
  36:17  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  37:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  38:5   error  Expected blank line before this statement                                                                                                                              padding-line-between-statements
  38:18  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  39:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  39:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  40:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  40:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  41:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  41:34  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  42:37  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  43:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  43:6   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  44:1   error  Trailing spaces not allowed                                                                                                                                            no-trailing-spaces
  44:5   error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  45:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  45:30  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  46:13  error  Unexpected use of '^'                                                                                                                                                  no-bitwise
  46:30  error  Expected linebreaks to be 'LF' but found 'CRLF'                                                                                                                        linebreak-style
  47:1   error  Expected indentation of 0 spaces but found 4                                                                                                                           indent
  47:6   error  Newline required at end of file but not found                                                                                                                          eol-last

✖ 93 problems (93 errors, 0 warnings)
  83 errors and 0 warnings potentially fixable with the `--fix` option.

File: index.js [hunk 1]
```diff
@@ -1,11 +1,39 @@
...
-
-module.exports = {
-    addNumber
-};
\ No newline at end of file
+function isNotANumber(item) {
+    return typeof item !== 'number';
+}
+
+function isSomeArrayValueNaN(item) {
```
Comment (ghost) [inline index.js]: Вообще, есть метод isNaN, но у него внутренняя логика отличается от вашей, поэтому некорректно называть так функцию, если она проверяет не совсем то, что нужно.

Comment (AlexNg9) [reply index.js]: И эта функция даже не используется.

Comment (ghost) [reply index.js]: И какой смысл пробрасывать аргументы в другую функцию? Можно убрать эту.

File: index.js [hunk 2]
```diff
@@ -1,11 +1,39 @@
...
+    return isNotANumber(item);
+}
+
+function isNumberPositiveAndInteger(num) {
+    if (!isNotANumber(num) && Number.isInteger(num) && (num >= 0)) {
```
Comment (ghost) [inline index.js]: Проверка на число избыточна, если есть isInteger.

Comment (ghost) [reply index.js]: Можно сразу возвращать это выражение, так как оно принимает значение true/false.

File: index.js [hunk 3]
```diff
@@ -1,11 +1,39 @@
...
+    return false;
+}
+
+function addNumber(array, number) {
+    if (array.some(isSomeArrayValueNaN)) {
+        return null;
+    }
+
+    if (!isNumberPositiveAndInteger(number)) {
+        return false;
```
Comment (ghost) [inline index.js]: Функция не должна возвращать false. Массив или null.
Два условия можно объединить в одно.

File: index.js [hunk 4]
```diff
@@ -1,11 +1,39 @@
...
+    }
+
+    let firstTerm = '';
+    array.forEach((item) => {
```
Comment (ghost) [inline index.js]: Посмотрите на reduce.

**Review by ghost — CHANGES_REQUESTED** [review]

File: index.js [hunk 5]
```diff
@@ -1,32 +1,21 @@
...
 }
 
-function isSomeArrayValueNaN(item) {
-    return isNotANumber(item);
-}
-
 function isNumberPositiveAndInteger(num) {
-    if (!isNotANumber(num) && Number.isInteger(num) && (num >= 0)) {
-        return true;
+    if (num >= 0) {
```
Comment (ghost) [inline index.js]: Можно всю логику этой функции в одну строку переписать.

File: index.js [hunk 6]
```diff
@@ -1,32 +1,21 @@
...
 
-    if (!isNumberPositiveAndInteger(number)) {
-        return false;
-    }
-
-    let firstTerm = '';
-    array.forEach((item) => {
-        firstTerm += item.toString();
-    });
+    let firstTerm = array.reduce((outputStr, curItem) => outputStr += curItem.toString());
```
Comment (ghost) [inline index.js]: Лучше в reduce всегда явно передавать начальное значение (аккумулятор).
И нет смысл делать присвоение `+=` – можно возвращать сумму.

Comment (ghost) [inline index.js]: Не понимаю смысла названия переменной.

**Review by ghost — CHANGES_REQUESTED** [review]

File: index.js [hunk 7]
```diff
@@ -1,11 +1,23 @@
...
-module.exports = {
-    addNumber
-};
\ No newline at end of file
+function checkTheNumberType(item) {
+    return typeof item !== 'number';
+}
+
+function isNumberPositiveAndInteger(num) {
+    return num >= 0 ? Number.isInteger(num) : false;
```
Comment (ghost) [inline index.js:6]: Число является положительным и целым, если оно положительное **и** (&&) целое.

**Review by ghost — APPROVED** [review]
👍

**Pull request closed without merging**